client/extensions/woocommerce/app/promotions/promotion-update.js

```diff
@@ -6,36 +6,211 @@
 
 import React from 'react';
 import PropTypes from 'prop-types';
+import { bindActionCreators } from 'redux';
 import { connect } from 'react-redux';
 import { localize } from 'i18n-calypso';
+import page from 'page';
+import { debounce } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import Main from 'components/main';
+import accept from 'lib/accept';
+import { successNotice, errorNotice } from 'state/notices/actions';
+import { getLink } from 'woocommerce/lib/nav-utils';
+import {
+	fetchPromotions,
+	updatePromotion,
+	deletePromotion,
+} from 'woocommerce/state/sites/promotions/actions';
+import { fetchProductCategories } from 'woocommerce/state/sites/product-categories/actions';
+import { getProductCategories } from 'woocommerce/state/sites/product-categories/selectors';
+import { editPromotion, clearPromotionEdits } from 'woocommerce/state/ui/promotions/actions';
 import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import { fetchSettingsGeneral } from 'woocommerce/state/sites/settings/general/actions';
+import { getPaymentCurrencySettings } from 'woocommerce/state/sites/settings/general/selectors';
+import {
+	getPromotionWithLocalEdits,
+	getPromotionableProducts,
+} from 'woocommerce/state/selectors/promotions';
+import { isValidPromotion } from './helpers';
+import PromotionHeader from './promotion-header';
+import PromotionForm from './promotion-form';
 
 class PromotionUpdate extends React.Component {
 	static propTypes = {
+		className: PropTypes.string,
 		site: PropTypes.shape( {
 			ID: PropTypes.number,
+			slug: PropTypes.string,
 		} ),
-		className: PropTypes.string,
+		promotion: PropTypes.shape( {
+			id: PropTypes.isRequired,
+		} ),
+		editPromotion: PropTypes.func.isRequired,
+		clearPromotionEdits: PropTypes.func.isRequired,
+		fetchSettingsGeneral: PropTypes.func.isRequired,
+		fetchPromotions: PropTypes.func.isRequired,
+		fetchProductCategories: PropTypes.func.isRequired,
+		updatePromotion: PropTypes.func.isRequired,
+		deletePromotion: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			busy: false,
+		};
+	}
+
+	componentDidMount() {
+		const { site } = this.props;
+
+		if ( site && site.ID ) {
+			this.props.fetchProductCategories( site.ID );
+			this.props.fetchPromotions( site.ID );
+			this.props.fetchSettingsGeneral( site.ID );
+		}
+	}
+
+	componentWillReceiveProps( newProps ) {
+		const { site } = this.props;
+		const newSiteId = ( newProps.site && newProps.site.ID ) || null;
+		const oldSiteId = ( site && site.ID ) || null;
+		if ( oldSiteId !== newSiteId ) {
```

Comment:
should we verify that `newSiteId` is not `null` here before calling the query components?

Reply:
Hmm, that's a fair point. However, in practice, this doesn't ever happen. Nearly exact copies of this code exist so many places in this app. Honestly, I'd like to do something a bit more elegant for all of them.

Reply:
Fair enough, or just ensure the query components bail if `siteId` is null then not worry about it 👍